canon/src/store/mod.rs

```diff
@@ -0,0 +1,138 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use cfg_if::cfg_if;
+
+use crate::{Canon, CanonError, Id};
+
+use alloc::vec::Vec;
+
+cfg_if! {
+    if #[cfg(feature = "host")] {
+        mod host;
+    } else {
+        mod hosted;
+    }
+}
+
+/// The singleton used to access the current store
+pub struct Store;
+
+impl Store {
+    /// Fetch bytes of an Id into the specified buffer
+    pub fn fetch(id: &Id, into: &mut [u8]) -> Result<(), CanonError> {
+        cfg_if! {
+            if #[cfg(feature = "host")] {
+                host::HostStore::fetch(id, into)
+            } else {
+                hosted::BridgeStore::fetch(id, into)
+            }
+        }
+    }
+
+    /// Get a value from storage, given an identifier
+    pub fn get<T: Canon>(id: &Id) -> Result<T, CanonError> {
+        cfg_if! {
+            if #[cfg(feature = "host")] {
+                host::HostStore::get(id)
+            } else {
+                hosted::BridgeStore::get(id)
+            }
+        }
+    }
+
+    /// Encode a value into the store
+    pub fn put<T: Canon>(t: &T) -> Id {
+        cfg_if! {
+            if #[cfg(feature = "host")] {
+                host::HostStore::put(t)
+            } else {
+                hosted::BridgeStore::put(t)
+            }
+        }
+    }
+
+    /// Get the id of a type, without storing it
+    pub fn id<T: Canon>(t: &T) -> Id {
+        cfg_if! {
+            if #[cfg(feature = "host")] {
+                host::HostStore::id(t)
+            } else {
+                hosted::BridgeStore::id(t)
+            }
+        }
+    }
+
+    /// Hash a slice of bytes
+    pub fn hash(bytes: &[u8]) -> [u8; 32] {
+        cfg_if! {
+            if #[cfg(feature = "host")] {
+                host::HostStore::hash(bytes)
+            } else {
+                hosted::BridgeStore::hash(bytes)
+            }
+        }
+    }
+
+    /// Hash a type implementing Canon.
+    pub fn canon_hash<T: Canon>(t: &T) -> [u8; 32] {
+        let len = t.encoded_len();
+        let mut buf = Vec::with_capacity(len);
+        buf.resize_with(len, || 0);
+        let mut sink = Sink::new(&mut buf[..]);
+        t.write(&mut sink);
+        Self::hash(&buf[..])
+    }
+}
```

Comment:
I think we should have a `Store` implementation in `host` and a `Store` implementation in `hosted` and the first `if` (lines 13-19) dictate which to use. Common implementation can stay here (such as `canon_hash`).
We did something similar for `rusk_abi`.

Reply:
I agree, i reorganized to only use _one_ #cfg_if